From 0c59f893c6f07c04e4d54aec73aa0e38dbf58de6 Mon Sep 17 00:00:00 2001
From: Nirmal Raj S <nirm@qti.qualcomm.com>
Date: Mon, 6 Apr 2026 10:59:21 +0530
Subject: [PATCH] handle sta connected event in all states

Handle sta connect & disconnect event to accomodate link addr

Removed 5sec timeout in fsm_setup for KPI optimization

Signed-off-by: Nirmal Raj S <nirm@qti.qualcomm.com>
---
 .../backhaul_manager/backhaul_manager.cpp     | 11 ++++-
 agent/src/beerocks/slave/son_slave_thread.cpp | 42 ++++++++++++++++---
 .../bwl/nl80211/base_wlan_hal_nl80211.cpp     | 11 +----
 3 files changed, 46 insertions(+), 18 deletions(-)

diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
index b7c0a5869..4fba447cb 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
@@ -2460,8 +2460,7 @@ bool BackhaulManager::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t even
             // Send slave enable the AP's
             send_slaves_enable();
             FSM_MOVE_STATE(CONNECTED);
-        }
-        if (FSM_IS_IN_STATE(WIRELESS_ASSOCIATE_4ADDR_WAIT)) {
+        } else if (FSM_IS_IN_STATE(WIRELESS_ASSOCIATE_4ADDR_WAIT)) {
             LOG(DEBUG) << "successful connect on iface=" << iface;
             if (hidden_ssid) {
                 iface_hal->refresh_radio_info();
@@ -2524,6 +2523,14 @@ bool BackhaulManager::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t even
                 m_task_pool.send_event(eTaskType::TOPOLOGY, TopologyTask::eEvent::AGENT_RADIO_STATE_CHANGED);
             }
 
+            FSM_MOVE_STATE(CONNECTED);
+        } else if ((FSM_IS_IN_STATE(INITIATE_SCAN)) || (FSM_IS_IN_STATE(WAIT_FOR_SCAN_RESULTS)) || (FSM_IS_IN_STATE(WIRELESS_CONFIG_4ADDR_MODE))){
+            LOG(DEBUG) << "Received STA Connected event in unexpected state -- move to CONNECTED state";
+            roam_flag      = false;
+            state_attempts = 0;
+            // Send slaves to enable the AP's
+            send_slaves_enable();
+
             FSM_MOVE_STATE(CONNECTED);
         }
     } break;
diff --git a/agent/src/beerocks/slave/son_slave_thread.cpp b/agent/src/beerocks/slave/son_slave_thread.cpp
index 19b3ddd75..0fd192f40 100644
--- a/agent/src/beerocks/slave/son_slave_thread.cpp
+++ b/agent/src/beerocks/slave/son_slave_thread.cpp
@@ -2661,15 +2661,24 @@ bool slave_thread::process_client_association(
         // as MLO clients maintain their association across multiple links.
         db->erase_client(client_mac);
 
+        // Parse link_mac from the notification message
+        auto &link_mac = notification_in->link_mac();
+
+#ifndef USE_PRPLMESH_NLHAL
         // Get radio for the legacy client using BSSID
         auto radio = db->get_radio_by_mac(bssid, AgentDB::eMacType::BSSID);
         if (!radio) {
             LOG(DEBUG) << "Radio containing bssid " << bssid << " not found";
             return false;
         }
-
-        // Parse link_mac from the notification message
-        auto &link_mac = notification_in->link_mac();
+#else
+        // Get radio for the legacy client using link_mac
+        auto radio = db->get_radio_by_mac(link_mac, AgentDB::eMacType::BSSID);
+        if (!radio) {
+            LOG(DEBUG) << "Radio containing bssid " << bssid << " not found";
+            return false;
+        }
+#endif
 
         radio->associated_clients.emplace(
             client_mac, AgentDB::sRadio::sClient{bssid, notification_in->association_frame_length(),
@@ -3127,7 +3136,7 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
         }
 
         // If exists, remove client association information for disconnected client.
-        db->erase_client(client_mac, bssid);
+        db->erase_client(client_mac);
 
         if (reconcile_ts_sta_ifaces) {
             LOG(DEBUG) << "Trigger traffic separation WDS clear on backhaul-BSS disconnect";
@@ -3442,8 +3451,10 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
         LOG(TRACE) << "received ACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION";
         auto &client_mac = notification_in->mac();
         auto &bssid      = notification_in->bssid();
-        LOG(INFO) << "Client associated sta_mac=" << client_mac << " to bssid=" << bssid
-                  << " is_mlo=" << notification_in->is_mlo();
+
+        auto &link_mac = notification_in->link_mac();
+
+        LOG(DEBUG) << "Client associated sta_mac=" << client_mac << " to bssid=" << bssid << " to linkmac=" << link_mac;
 
         // Check if the client is an Multi-AP Agent, '0' means a regular station.
         if (notification_in->multi_ap_profile() != 0) {
@@ -3455,6 +3466,8 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
         // Save information AgentDB
         auto db = AgentDB::get();
 
+
+#ifndef USE_PRPLMESH_NLHAL
         // Set client association information for associated client
         auto radio = db->get_radio_by_mac(bssid, AgentDB::eMacType::BSSID);
         if (!radio) {
@@ -3469,6 +3482,23 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
             }
         }
 
+#else
+        // Set client association information for associated client
+        auto radio = db->get_radio_by_mac(link_mac, AgentDB::eMacType::BSSID);
+        if (!radio) {
+            LOG(DEBUG) << "Radio containing bssid " << bssid << " not found";
+            break;
+        }
+
+        for (auto &bss : radio->front.bssids) {
+            if (bss.mac == link_mac && bss.backhaul_bss) {
+                m_task_pool.send_event(eTaskType::AP_AUTOCONFIGURATION,
+                                       ApAutoConfigurationTask::eEvent::APPLY_CONFIG_FOR_NEW_IFACE);
+            }
+        }
+
+#endif
+
         if (!process_client_association(notification_in)) {
             LOG(WARNING) << "Failed to process client association for " << client_mac;
         }
diff --git a/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp
index 6a820ca5e..fef07158c 100644
--- a/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp
@@ -240,17 +240,8 @@ bool base_wlan_hal_nl80211::fsm_setup()
 
         .state(nl80211_fsm_state::Delay)
 
-        // On Entry
-        .entry([&](const void *args) -> bool {
-            m_state_timeout = std::chrono::steady_clock::now() + std::chrono::seconds(5);
-            return true;
-        })
-
         // EVENT -> Attach
-        .on(nl80211_fsm_event::Attach, nl80211_fsm_state::Init,
-            [&](TTransition &transition, const void *args) -> bool {
-                return (std::chrono::steady_clock::now() >= m_state_timeout);
-            })
+        .on(nl80211_fsm_event::Attach, nl80211_fsm_state::Init)
 
         //////////////////////////////////////////////////////////////////////////
         ////////////////////////////// State: Init ///////////////////////////////
-- 
2.34.1

